pkg/evs/controllerserver.go

```diff
@@ -264,20 +267,189 @@ func (cs *ControllerServer) ListVolumes(_ context.Context, req *csi.ListVolumesR
+	credentials := cs.Driver.cloudCredentials
+	name := req.GetName()
+	volumeId := req.GetSourceVolumeId()
+	if err := checkCreateSnapshotParamEmpty(name, volumeId); err != nil {
+		return nil, err
+	}
+	if response, err := checkRepeatSnapshotName(credentials, name, volumeId); response != nil || err != nil {
+		return response, err
+	}
+	if err := checkVolumeIsExist(credentials, volumeId); err != nil {
+		return nil, err
+	}
+	return doSnapshotCreate(credentials, name, volumeId)
 }
 
-func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (
-	*csi.DeleteSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func checkCreateSnapshotParamEmpty(name string, volumeId string) error {
+	if volumeId == "" {
+		return status.Error(codes.InvalidArgument, "CreateSnapshot SourceVolumeId cannot be empty")
+	}
+	if name == "" {
+		return status.Error(codes.InvalidArgument, "CreateSnapshot Name cannot be empty")
+	}
+	return nil
 }
 
-func (cs *ControllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (
-	*csi.ListSnapshotsResponse,
-	error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func checkRepeatSnapshotName(credentials *config.CloudCredentials, name string, volumeId string) (*csi.CreateSnapshotResponse, error) {
+	listOpts := snapshots.ListOpts{
+		Name: name,
+	}
+	listSnapshots, err := services.ListSnapshots(credentials, listOpts)
+	if err != nil {
+		klog.Errorf("Failed to query snapshot list: %v", err)
+		return nil, status.Error(codes.Internal, "Failed to get snapshots")
+	}
+	if len(listSnapshots) == 1 {
+		snap := &listSnapshots[0]
+		if snap.VolumeID != volumeId {
+			// same name with different volumeId
+			return nil, status.Error(codes.AlreadyExists, "Snapshot with given name already exists, with different source volume ID")
+		}
+		klog.V(3).Infof("Found existing snapshot %s on %s", name, volumeId)
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				SnapshotId:     snap.ID,
+				SizeBytes:      int64(snap.Size * 1024 * 1024 * 1024),
+				SourceVolumeId: snap.VolumeID,
+				CreationTime:   timestamppb.New(snap.CreatedAt),
+				ReadyToUse:     true,
+			},
+		}, nil
+	}
+	if len(listSnapshots) > 1 {
+		// many repeat name
+		klog.Errorf("Found multiple existing snapshots with selected name (%s) during create", name)
+		return nil, status.Error(codes.Internal, "Multiple snapshots reported by Cinder with same name")
+	}
+	return nil, nil
+}
+
+func checkVolumeIsExist(credentials *config.CloudCredentials, volumeId string) error {
+	volume, err := services.GetVolume(credentials, volumeId)
+	if err != nil {
+		klog.Errorf("Failed to query volume: %v", err)
+		return err
+	}
+	if volume == nil {
+		return status.Error(codes.Internal, "Volume id not exist")
+	}
+	return nil
+}
+
+func doSnapshotCreate(credentials *config.CloudCredentials, name string, volumeId string) (*csi.CreateSnapshotResponse, error) {
```

Comment:
`doSnapshotCreate` looks like a service, move it to the services pakcage and name it `CreateSnapshotToCompleted`